### ISSUE TITLE
About the GroupNorm implementation

### ISSUE THREAD

**Issue by GlassyWing** [pytorch/pytorch#87846]
I find that after applying `GroupNorm` (pytorch offical implementation) in my network, the effect will be significantly reduced. This happens in almost all types of networks, such as VAE, DDPM, GAN, and target detection. In particular, in a generative type of network, the final generated image is clearly more like an average of multiple batches of images. 

Previously I thought it was a problem with the method itself, until I implemented a version myself based on the GroupNorm documentation in the torch.

```
class GroupNormCustom(nn.Module):

    def __init__(self, n_groups, num_channels, eps=1e-6, affine=True):
        super().__init__()
        self.n_groups = n_groups
        self.num_channels = num_channels
        self.eps = eps
        self.affine = affine
        if self.affine:
            self.weight = Parameter(torch.empty(num_channels))
            self.bias = Parameter(torch.empty(num_channels))
        else:
            self.register_parameter('weight', None)
            self.register_parameter('bias', None)

        self.reset_parameters()

    def reset_parameters(self) -> None:
        if self.affine:
            init.ones_(self.weight)
            init.zeros_(self.bias)

    def forward(self, x):
        b, c, h, w = x.shape
        x = x.permute(0, 2, 3, 1)
        x = x.reshape(b, h, w, self.n_groups, c // self.n_groups)  # (b, h, w, g, f) s.t. c = g * f
        var, mean = torch.var_mean(x, dim=[1, 2, 3], keepdim=True)
        # norm_x = (x - mean) * torch.rsqrt(var + self.eps)  # (b, h, w, g, f)
        norm_x = x.sub(mean).mul(torch.rsqrt(var + self.eps))
        norm_x = norm_x.flatten(start_dim=-2)  # (b, h, w, c)
        if self.affine:
            # norm_x = norm_x * self.weight + self.bias
            norm_x = norm_x.mul(self.weight[None, None, None]).add(self.bias[None, None, None])
        return norm_x.permute(0, 3, 1, 2)
```

It brings significant performance improvements. I don't think my implementation deviates from the official documentation, but the result is much better than the official implementation.

Since I have no experience with cuda programming, I don't know the details of the official implementation, can you please explain what it does using python style code?

**Comment by helson73** [comment]
Did you use `memory_format=channels_last` in your data and networks ?

**Comment by GlassyWing** [comment]
> 

I don't know what that is, so I'm sure it's not used.

**Comment by GlassyWing** [comment]
> Did you use `memory_format=channels_last` in your data and networks ?

I don't know what that is, so I'm sure it's not used.

**Comment by GlassyWing** [comment]
Now I have know what happening in torch's implementation, it flatten all dims to shape (B, C x H x W), and then do grouping at 2nd dim, It is actually operating on spatially non-adjacent pixels, This is the completely wrong way, so the effect is not good. And this also causes some details in the diffusion model to be generated incorrectly, such as the distortion of the hand.
